packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart

```diff
@@ -2,18 +2,119 @@
+class VsCodeFlutterPanel extends StatefulWidget {
+  const VsCodeFlutterPanel(this.api, {super.key});
+
+  final DartToolingApi api;
 
-class VsCodeFlutterPanel extends StatelessWidget {
-  const VsCodeFlutterPanel({super.key});
+  @override
+  State<VsCodeFlutterPanel> createState() => _VsCodeFlutterPanelState();
+}
 
+class _VsCodeFlutterPanelState extends State<VsCodeFlutterPanel> {
   @override
   Widget build(BuildContext context) {
     assert(FeatureFlags.vsCodeSidebarTooling);
-    return const Center(
-      child: Text('TODO: a panel for flutter actions in VS Code'),
+
+    final api = widget.api;
+
+    return Column(
+      children: [
+        const Text(''),
+        FutureBuilder(
+          future: api.vsCode,
+          builder: (context, snapshot) =>
+              switch ((snapshot.connectionState, snapshot.data)) {
+            (ConnectionState.done, final vsCodeApi?) =>
+              _VsCodeConnectedPanel(vsCodeApi),
+            (ConnectionState.done, null) =>
+              const Text('VS Code is not available'),
+            _ => const CenteredCircularProgressIndicator(),
+          },
+        ),
+      ],
+    );
+  }
+}
+
+/// The panel shown once we know VS Code is available (the host has responded to
+/// the `vsCode.getCapabilities` request).
+class _VsCodeConnectedPanel extends StatefulWidget {
+  const _VsCodeConnectedPanel(this.api, {super.key});
+
+  final VsCodeApi api;
+
+  @override
+  State<_VsCodeConnectedPanel> createState() => _VsCodeConnectedPanelState();
+}
+
+class _VsCodeConnectedPanelState extends State<_VsCodeConnectedPanel> {
+  @override
+  void initState() {
+    super.initState();
+
+    unawaited(widget.api.initialize());
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: [
+        if (widget.api.capabilities.executeCommand)
+          ElevatedButton(
+            onPressed: () =>
+                unawaited(widget.api.executeCommand('flutter.createProject')),
+            child: const Text('New Flutter Project'),
+          ),
+        StreamBuilder(
+          stream: widget.api.devicesChanged,
+          builder: (context, snapshot) {
+            if (!snapshot.hasData) {
+              return const Text('');
+            }
+            final deviceEvent = snapshot.data!;
+            return Table(
+              defaultVerticalAlignment: TableCellVerticalAlignment.middle,
+              children: [
+                for (final device in deviceEvent.devices)
+                  TableRow(
+                    children: [
+                      TextButton(
+                        child: Text(device.name),
+                        onPressed: () =>
+                            unawaited(widget.api.selectDevice(device.id)),
+                      ),
+                      Text(
+                        device.id == deviceEvent.selectedDeviceId
+                            ? '(selected)'
+                            : '',
+                      ),
```

Comment:
this text can be part of the TextButton child Text widget above.

```
Text(
  '${device.name}'
  '${device.id == deviceEvent.selectedDeviceId ? ' (selected)' : ''}',
)
```

Reply:
That would include it in the button so the styling would change - this was deliberately separate:

![image](https://github.com/flutter/devtools/assets/1078012/382e3176-499c-49ed-8c34-4e1fe931b29b)

Although, this "design" is just a placeholder and we definitely don't want to ship it looking like this (we probably want to show more than just the name of a device for ex.).

Reply:
Fine to land as is and iterate with some ux input later